### DISCUSSION TITLE
twinBASIC BETA 72

### DISCUSSION THREAD

**Discussion by WaynePhillipsEA** [twinbasic/twinbasic#981] [Releases]
- fixed: unary operator detection involving Case and line labels was sometimes incorrect [ https://github.com/twinbasic/twinbasic/issues/969#issuecomment-1179674485 ]
- fixed: built-in enumeration priority order (...vbNormal is ambiguous conflict) [ https://github.com/twinbasic/twinbasic/issues/718 ]
- added: Globals.Unload() support  [ https://github.com/twinbasic/twinbasic/issues/783 ]
- improved: overhauled the form Load/Unload routines to better match VBx
- fixed: Form.Show now offers the Modal argument.  Modeless is not yet supported, so vbModal must now be explicitly passed here.

<hr /><em>This discussion was created from the release <a href='https://github.com/twinbasic/twinbasic/releases/tag/beta-72'>twinBASIC BETA 72</a>.</em>

**Comment by WaynePhillipsEA** [comment]
Apologies as the change to `Form.Show` is a breaking change in tB, but of course needed for backwards compatibility with VBx.  I should've added the parameter from the start regardless of not supporting modeless forms.  If you now just call `Form.Show()`, it will raise an error as we don't yet support modeless forms.  Change all uses to `Form.Show(vbModal)` to fix it.

In addition, I'm expecting a few regressions in this release, since there were significant changes to the form Load/Unload routines to match VBx.  But the compatibility is improved, so for instance you can now Unload a form and reuse the same form instance, just like in VB6.

**Reply by bclothier** [reply]
Don't know if that's applicable but `UserForm1.Show` for a VBA userform is legal as the argument is optional. If omitted, it's shown modally. You'd need to specify the argument for modeless.

**Reply by WaynePhillipsEA** [reply]
The argument is optional in VB6 forms also, but the default is modeless, not modal.

**Reply by bclothier** [reply]
Oof. That's going to create a headache when support for importing a VBA project with userforms is enabled....

I guess one way to smooth over that is to convert all `.Show()` from VB6 project into `.Show(vbModeless)` and for VBA, `Show()` into `.Show(vbModal)`. Not 100% ideal since this changes the source code but at least it'd behave the same way regardless where it came from.

**Reply by WaynePhillipsEA** [reply]
TBH we'll probably need a separate Form type when it comes to UserForms.  So the UserForm implementation will be able to keep its modal-default Show() implementation.

**Reply by EduardoVB** [reply]
> Oof. That's going to create a headache when support for importing a VBA project with userforms is enabled....

I've been thinking about VBA forms, and they seem to be quite different from VB6 forms in some regards.
The events are not the same.
I didn't study the issue much, I just was checking some things in a Word UserForm.
And I was thinking in what way tB could support both, without overlapping features. 

Because I think the only way to have full backward compatibility with both environments would be to add a type UserForm, but that would go against the rule of "only one only way to do something".

Edit: I didn't read this when I wrote the message:

> TBH we'll probably need a separate Form type

**Comment by OldestCoderInTown** [comment]
I apologise for not picking up the need for a vbmodal bit being required, where I think have not manually typed a form.show, but I will look into fixing that. I have tested my program with every release since 65 to 73 and only 72 & 73 fail. I did notice that 71 is also called 70 which caused me some head scratching during the testing.

**Reply by OldestCoderInTown** [reply]
Is this the offending bit of code?

Public Sub Show(Optional ByVal Modal As FormShowConstants)
        If Modal = FormShowConstants.vbModeless Then
            Err.Raise 5, , "Sorry, twinBASIC doesn't yet support non-modal forms.  To show a modal form use Show(vbModal)"
        End If
        Me.ControlContext.EnsureFormIsLoaded()
        **Me.FormControlContext.Show()**
    End Sub

**Reply by OldestCoderInTown** [reply]
Is this the bit that needs changed, and if so, how do I edit the baseform code?

Public Sub Show()
        Me.FormControlContext.Show()
    End Sub


My attempts to edit that are met with "Cannot edit in read-only editor"
